lib/Dialect/LLHD/IR/LLHDOps.cpp

```diff
@@ -388,6 +388,66 @@ OpFoldResult llhd::ShrOp::fold(ArrayRef<Attribute> operands) {
       });
 }
 
+//===----------------------------------------------------------------------===//
+// ExtractElementOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult llhd::ExtractElementOp::fold(ArrayRef<Attribute> operands) {
+  uint64_t index = indexAttr().getInt();
+
+  // llhd.extract_element(llhd.shr(hidden, base, constant amt), index)
+  if (auto shrOp = target().getDefiningOp<llhd::ShrOp>()) {
+    IntegerAttr intAttr;
+    if (matchPattern(shrOp.amount(), m_Constant<IntegerAttr>(&intAttr))) {
```

Comment:
would it make sense to merge the `getDefiningOp` check into the pattern match?

Reply:
To my knowledge there does not exist a matcher that allows to bind the `shrOp` or its operands to values as the `m_Constant` does for attributes. This would make it very simple and clean. Let me know if I overlooked something!
But I can merge it in the `matchPattern` and then extract the `shrOp` manually within the if statement to get rid of one.

Reply:
I'm pretty sure you can use `m_Op<llhd::ShrOp>(..)` for this, but I'm not completely sure.